// Copyright 2010-2012, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto2";

package mozc.user_dictionary;

message UserDictionaryStorage {
  message UserDictionary {
    // ID of this dictionary
    optional uint64 id = 1 [ default = 0 ];

    // set false if this dictionary is not used.
    // Even if |enabled| is false, the dictionary
    // it self is visible to user.
    optional bool enabled = 2 [ default = true ];

    // name of dictionary
    optional string name = 3 [ default = "" ];

    // entry of each word
    message Entry {
      optional string key     = 1 [ default = "" ];
      optional string value   = 2 [ default = "" ];
      optional string pos     = 3 [ default = "" ];
      optional string comment = 4 [ default = "" ];

      // set true if this entry is removed.
      // This flag is used for cloud sync feature.
      optional bool  removed  = 10 [ default = false ];
    };

    repeated Entry entries = 4;

    // set true if this dictionary is removed.
    // This flag is used for cloud sync feature.
    optional bool removed = 5 [ default = false ];

    // This flag is used for cloud sync feature.
    optional bool syncable = 6 [ default = false ];
  };

  // version of user dictionary
  optional int32 version = 1 [ default = 0 ];

  // dictionary body
  repeated UserDictionary dictionaries = 2;

};
